tiny_dnn/core/kernels/global_avepool_grad_op.h

```diff
@@ -27,10 +28,28 @@ class GlobalAvePoolGradOp : public core::OpKernel {
     // initialize outputs
     fill_tensor(prev_delta, float_t{0});
 
-    // only internal kernel op implemented yet, so use it regardless
-    // of the specified backend engine
-    kernels::global_avepool_grad_op_internal(prev_delta, curr_delta, params,
-                                             context.parallelize());
+    const core::backend_t engine = context.engine();
+
+    if (engine == core::backend_t::internal) {
+      kernels::global_avepool_grad_op_internal(prev_delta, curr_delta, params,
+                                               context.parallelize());
+    } else if (engine == core::backend_t::avx) {
+#ifdef CNN_USE_DOUBLE
+      // todo (kd): add avx implementation for CNN_USE_DOUBLE
+      kernels::global_avepool_grad_op_internal(prev_delta, curr_delta, params,
+                                               context.parallelize());
+#else
+      kernels::global_avepool_grad_op_avx(prev_delta, curr_delta, params,
+                                          context.parallelize());
+#endif
+    } else if (engine == core::backend_t::nnpack) {
+      // fallback to internal implementation as nnpack implementation
+      // is not available
+      kernels::global_avepool_grad_op_internal(prev_delta, curr_delta, params,
+                                               context.parallelize());
+    } else {
+      throw nn_error("Not supported engine: " + to_string(engine));
+    }
```

Comment:
I don't think all backends should ready its own versions of every kernels as it's highly likely impossible.
so I would simply write fallback routine in this way.
```cpp
    if (engine == core::backend_t::avx) {
#ifdef CNN_USE_DOUBLE
      // todo (kd): add avx implementation for CNN_USE_DOUBLE
      kernels::global_avepool_grad_op_internal(prev_delta, curr_delta, params,
                                               context.parallelize());
#else
      kernels::global_avepool_grad_op_avx(prev_delta, curr_delta, params,
                                          context.parallelize());
#endif
    } else {
      kernels::global_avepool_grad_op_internal(prev_delta, curr_delta, params,
                                               context.parallelize());
    }
```